src/rpc_client/src/meta_client.rs

```diff
@@ -851,93 +868,335 @@ impl HummockMetaClient for MetaClient {
+                // Risedev: May happen if there is a problem with Nginx routing the requests
+                failure_msgs
+                    .push("Unable to retrieve leader address via service. Retrying...".to_owned());
+                tokio::time::sleep(retry).await;
+                continue;
+            }
+            let current_leader = nl.unwrap();
+
+            let addr = format!(
+                "http://{}:{}",
+                current_leader.get_host(),
+                current_leader.get_port()
+            );
+            let leader_channel = match get_channel_no_retry(addr.as_str()).await {
+                Ok(c) => c,
+                Err(e) => {
+                    failure_msgs.push(format!( 
+                         "Failed to establish connection with {}. Err: {}. Assuming stale leader info. Retrying...",
+                        addr, e
+                    ));
+                    tokio::time::sleep(retry).await;
+                    continue;
+                }
+            };
+
+            // Hold locks on all sub-clients, to update atomically
+            {
+                let mut leader_c = self.leader_client.as_ref().lock().await;
+                let mut cluster_c = self.cluster_client.as_ref().lock().await;
+                let mut heartbeat_c = self.heartbeat_client.as_ref().lock().await;
+                let mut ddl_c = self.ddl_client.as_ref().lock().await;
+                let mut hummock_c = self.hummock_client.as_ref().lock().await;
+                let mut notification_c = self.notification_client.as_ref().lock().await;
+                let mut stream_c = self.stream_client.as_ref().lock().await;
+                let mut user_c = self.user_client.as_ref().lock().await;
+                let mut scale_c = self.scale_client.as_ref().lock().await;
+                let mut backup_c = self.backup_client.as_ref().lock().await;
+
+                *leader_c = LeaderServiceClient::new(leader_channel.clone());
+                *cluster_c = ClusterServiceClient::new(leader_channel.clone());
+                *heartbeat_c = HeartbeatServiceClient::new(leader_channel.clone());
+                *ddl_c = DdlServiceClient::new(leader_channel.clone());
+                *hummock_c = HummockManagerServiceClient::new(leader_channel.clone());
+                *notification_c = NotificationServiceClient::new(leader_channel.clone());
+                *stream_c = StreamManagerServiceClient::new(leader_channel.clone());
+                *user_c = UserServiceClient::new(leader_channel.clone());
+                *scale_c = ScaleServiceClient::new(leader_channel.clone());
+                *backup_c = BackupServiceClient::new(leader_channel);
+            } // release MutexGuards on all clients
+
+            tracing::info!("Successfully failed over meta_client to {}", addr);
+            return;
+        }
+
+        // Only print failure messages if the entire failover failed
```

Comment:
I am doing this, to avoid spamming the logs. Printing immediately is very verbose, since the meta nodes send stale leader information for quite some time.